lib/pbench/server/cache_manager.py

```diff
@@ -194,6 +181,82 @@ def make_cache_object(dir_path: Path, path: Path) -> CacheObject:
     )
 
 
+class LockRef:
+    """Keep track of a cache lock passed off to a caller"""
+
+    def __init__(self, lock: Path, exclusive: bool = False, wait: bool = True):
+        """Initialize a lock reference
+
+        The lock file is opened in "w+" mode, which is "write update": unlike
+        "r+", this creates the file if it doesn't already exist, but still
+        allows lock conversion between LOCK_EX and LOCK_SH.
+
+        Args:
+            lock: the path of a lock file
+            exclusive: lock for exclusive access
+            wait: wait for lock
+        """
+        self.lock = lock.open("w+")
+        self.locked = False
+        self.exclusive = exclusive
+        self.unlock = True
+        self.wait = wait
+
+    def acquire(self) -> "LockRef":
+        """Acquire the lock"""
+
+        cmd = fcntl.LOCK_EX if self.exclusive else fcntl.LOCK_SH
+        if not self.wait:
+            cmd |= fcntl.LOCK_NB
+        fcntl.lockf(self.lock, cmd)
+        self.locked = True
+        return self
+
+    def release(self):
+        """Release the lock and close the lock file"""
+
+        exception = None
+        try:
+            fcntl.lockf(self.lock, fcntl.LOCK_UN)
+            self.lock.close()
+            self.locked = False
+            self.exclusive = False
+        except Exception as e:
+            exception = e
+        finally:
+            # Release our reference to the lock file so that the
+            # object can be reclaimed.
+            self.lock = None
+        if exception:
+            raise exception
```

Comment:
Why did you structure the code so that we escape the `try` block and then (re)raise the exception instead of just not catching it in the `try` block?  (I think the result is the same, but this code is more complex and potentially error-prone.)

Reply:
It's not really "error prone" in any meaningful way. In this case, now, the `finally` would have been enough, but I think the code "evolved" that way ... it made sense in an earlier life, and never revised itself.

Reply:
The error-prone-ness arises as the code continues to evolve.